Gcp/clouddns/dns policy

Cluel355 opened the pull request:
updated new helper policies

Cluel355 committed:
dns_policy

Cluel355 committed:
merge

Cluel355 committed:
updated policies using new helper

Cluel355 committed:
more policies

Cluel355 committed:
updated commit

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/auto_create_network/policy.rego:
what attribute is this? I cannot find "auto_create_network" anywhere in the resource json plan?

your resource folder shows dns_managed_zone, but the resource being evaluated is google_computer_network?

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/cloud_logging/policy.rego:
i cannot find the "cloud_logging" attribute being evaluated? it's not in your c.tf, and your json plan is missing. bring it with you so it's easier to review your policies.

in this policy you're whitelisting empty space ""? why?

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/compute_subnetwork/policy.rego:
i cannot find the "compute_subnetwork" attribute being evaluated? it seems you're trying to evaluate ""ip_cidr_range" instead?

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/container_cluster/policy.rego:
google_container_cluster is not an attribute; it is a seperate resource; it needs its own folder of policies.

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/directory_namespace/policy.rego:
different service entirely? are you doing google_dns_<RESOURCE> or some other service? here it's google_service_<RESOURCE>, earlier it was google_container_<RESOURCE> and google_compute_<RESOURCE>? what is your service?

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/managed_zone_private/policy.rego:
where is this attribute? and why are you whitelisting an empty list?

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/private_forwarding/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist.

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/private_gke/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist.

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/private_peering/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist. wrong attribute config in the policy.

jerryd10 commented on policies/gcp/clouddns/dns_managed_zone/service_directory/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist. wrong attribute config in the policy.

jerryd10 commented on policies/gcp/clouddns/dns_policy/alt_name_server/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist. wrong attribute config in the policy.

jerryd10 commented on policies/gcp/clouddns/dns_policy/enable_inbound_forwarding/policy.rego:
policy makes sense, but you should not be whitelisting empty space, you should be whitelisting "false".

jerryd10 commented on policies/gcp/clouddns/dns_policy/enable_logging/policy.rego:
good. from a security perspective, logging is good for casual analysis and forensics. please remove the comments.

jerryd10 commented on policies/gcp/clouddns/dns_policy/networks/policy.rego:
irrelevant and incomplete. the attribute being evaluated does not exist. wrong attribute config in the policy.

visal-dam reviewed (changes requested):
First couple of policies have nothing to do with the google_dns service's resources. 

Remove all comments.

You do not need to do policies for all attributes; rather, policies are context-based, and we use the attributes to gather and validate information according to our policy.

Most of policies evaluate attributes that does not exist. This tells me that you do not understand how to access certain values properly using the attributes. Redo the policies and rewatch the videos, please.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/auto_create_network/policy.rego:
not a dns resource, you can remove this policy.

visal-dam replied:
same as above

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/cloud_logging/policy.rego:
i think you're referencing the wrong attribute here.

visal-dam replied:
associated plan.json is missing so i have no point of reference, but from the included c.tf, shouldnt the attribute path be instead ["cloud_logging_config", 0, "enable_logging"]. the value is corrrect.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/compute_subnetwork/policy.rego:
not a dns resource, you can remove this policy.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/container_cluster/policy.rego:
not a dns resource, you can remove this policy.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/directory_namespace/policy.rego:
not a dns resource, you can remove this policy.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/managed_zone_basic/policy.rego:
you're referencing the wrong attribute here; should be dns_name rather than parent.

visal-dam replied:
Idk what this policy is supposed to do, as it is described to check the dns namespace but an irrelevant attribute is referenced

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/managed_zone_private/policy.rego:
you're referencing the wrong attribute here; should be visibility.

visal-dam replied:
"private_visibility_config", 0, "networks", 0 "network_url"] and values: [pde.coding_for_yotham.net] as you are now evaluating the private_visibility_config attribute rather than the visibility

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/private_forwarding/policy.rego:
you're referencing the wrong attribute here. to simulate private forwarding, you can check the ipv4_address of the targeted name servers in the forwarding configs such that they are whitelisted IP addresses. you can access them as ["forwarding_config", 0, "target_name_servers", 0, "ipv4_address"]

note that the last '0' gives you the first address; '1' would give u the second one. You can perhaps restrict to only have a set of IP addresses: for example to restrict to only two, have another condition that checks for/whitelists "null" when you try to access ["forwarding_config", 0, "target_name_servers", 2, "ipv4_address"]; if a third ip address is in place, it is non-compliant.

visal-dam replied:
attribute should be ["forwarding_config", 0, "target_name_servers", 0, "ipv4_address"], 
values: ["139.0.0.1"]

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/private_gke/policy.rego:
you're referencing the wrong attribute here. you can access what you want similar to what i wrote above.

visal-dam replied:
shouldn't it be ["private_visibility_config", 0, "gke_cluster"]? I see that you leave the values [] empty to mean that it can be configured later, but you should add some examples in the values [] array to demonstrate how those values added later would look like.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/private_peering/policy.rego:
wrong attribute access, and you're whitelisting a empty list, which is probably not what you intended.

visal-dam commented on policies/gcp/clouddns/dns_managed_zone/service_directory/policy.rego:
you're referencing the wrong attribute here. you can access what you want similar to what i wrote above.

visal-dam replied:
attribute should be ["service_directory_config", 0, "namespace", 0, "namespace_url"]

and you're whitelisting an empty list, so the policy prevents a service dir from being configured?

visal-dam commented on policies/gcp/clouddns/dns_policy/alt_name_server/policy.rego:
you're referencing the wrong attribute here. you can access what you want similar to what i wrote above.

visal-dam replied:
attribute should be:
"attribute_path" : ["alternative_name_server_config", 0, "target_name_servers", 0, "ipv4_address"]
values shouldn't be empty, unless you're trying to whitelist empty space/null to block any alt name server configs? but given what is described, you can add in a list of IPs:

values: ["1.1.1.1", "2.3.2.1", etc...]
whitelist

OR
values: ["1[.]1[.]1[.]*", [[1, 2, 3, 4, 5, 6, 7, 10]]] # this allows the addresses: 1.1.1.1, 1.1.1.2, 1.1.1.3, ..., 1.1.1.7 and 1.1.1.10 but not 1.1.1.8, 1.1.1.9 as 8, 9 are not in the pattern list
pattern whitelist

visal-dam commented on policies/gcp/clouddns/dns_policy/enable_inbound_forwarding/policy.rego:
good policy, but i think u meant to whitelist 'false' rather than an empty list.

visal-dam commented on policies/gcp/clouddns/dns_policy/enable_logging/policy.rego:
good policy. logging is essential for forensics and troubleshooting.

visal-dam commented on policies/gcp/clouddns/dns_policy/networks/policy.rego:
you're referencing the wrong attribute here. you can access what you want similar to what i wrote above.

visal-dam replied:
correct attribute referencing, but the values to be whitelisted are incorrect. here you want to only allow certain network urls, such as "http://love_writing_polices_for_yotham.net.id" (note the http is also not secure).  you can either:

1) hardcode whitelisting:
"values": ["https://love_writing_polices_for_yotham.net.id"],
"policy_type" : "whitelist" 

OR even better 2) patterns:
"values": ["https://*[.]*", [["love_writing_polices_for_yotham"], ["net.id"]]],
"policy_type" : "pattern whitelist"

visal-dam reviewed (changes requested):
please remove all the comments.

please remove policies that do not evaluate google_dns resources.

the way you've referenced the attributes are incorrect. have a look at the videos at how we access nested attributes.

mostly whitelisting empty values - please fill up these values lists.

Cluel355 commented:
Hi Visal,

Maybe this will help explain the way I did the policies. I worked through each resource and used each resource as I came across it if I thought it was relevant. This is a learning curve and we only had a week or 2 to update the repo, start writing rules and understand how it all works.

A few more weeks would have been great to help get more proficient at this. 🙂

Cheers Simon


________________________________
From: jerryd10 ***@***.***>
Sent: Wednesday, May 14, 2025 7:48 PM
To: Hardhat-Enterprises/Policy-Deployment-Engine ***@***.***>
Cc: SIMON RANKMORE ***@***.***>; Author ***@***.***>
Subject: Re: [Hardhat-Enterprises/Policy-Deployment-Engine] Gcp/clouddns/dns policy (PR #12)


@jerryd10 commented on this pull request.

________________________________

On policies/gcp/clouddns/dns_managed_zone/auto_create_network/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2088538868>:

what attribute is this? I cannot find "auto_create_network" anywhere in the resource json plan?

your resource folder shows dns_managed_zone, but the resource being evaluated is google_computer_network?

—
Reply to this email directly, view it on GitHub<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#pullrequestreview-2839519981>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BHJOKWXQCZUNRUKXPHAUWET26MGPTAVCNFSM6AAAAAB4PY67HOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQMZZGUYTSOJYGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Cluel355 committed:
updated 17/5

visal-dam reviewed (changes requested):
2nd review round.

Policies very much improved, but you need to configure the attribute paths correctly, and include some values too. Sure these values lists can be configured later but you should add example values of what these later vals can look like.

Iirevalent policies have been removed which is good. the current policies are very much security-related, mostly with allowed dest networks. Good on ya.

Cluel355 commented:
Hi Visal,

Thanks for this - this made it a lot clearer for me and helped me to fine-tune the policies.

I have deleted managed_zone_basic just because it seemed problematic.

Made the following changes as requested:


On policies/gcp/clouddns/dns_managed_zone/auto_create_network/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096945668>:
same as above
________________________________
On policies/gcp/clouddns/dns_managed_zone/cloud_logging/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096943739>:
associated plan.json is missing so i have no point of reference, but from the included c.tf, shouldnt the attribute path be instead ["cloud_logging_config", 0, "enable_logging"]. the value is corrrect.

Added json file and corrected
________________________________
On policies/gcp/clouddns/dns_managed_zone/managed_zone_basic/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096940620>:
Idk what this policy is supposed to do, as it is described to check the dns namespace but an irrelevant attribute is referenced

Removed this policy
________________________________
On policies/gcp/clouddns/dns_managed_zone/managed_zone_private/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096939228>:
"private_visibility_config", 0, "networks", 0 "network_url"] and values: [pde.coding_for_yotham.net] as you are now evaluating the private_visibility_config attribute rather than the visibility

Added values and updated path
________________________________
On policies/gcp/clouddns/dns_managed_zone/private_forwarding/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096936120>:
attribute should be ["forwarding_config", 0, "target_name_servers", 0, "ipv4_address"],
values: ["139.0.0.1"]

array value of target_name_servers is 1 - pls check if you want to confirm ( I could be wrong) - updated the rest of it
________________________________
On policies/gcp/clouddns/dns_managed_zone/private_gke/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096932060>:
shouldn't it be ["private_visibility_config", 0, "gke_cluster"]? I see that you leave the values [] empty to mean that it can be configured later, but you should add some examples in the values [] array to demonstrate how those values added later would look like.

added values
________________________________
On policies/gcp/clouddns/dns_managed_zone/service_directory/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096928577>:
attribute should be ["service_directory_config", 0, "namespace", 0, "namespace_url"]
and you're whitelisting an empty list, so the policy prevents a service dir from being configured?

Updated
________________________________
On policies/gcp/clouddns/dns_policy/alt_name_server/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096924941>:
attribute should be:
"attribute_path" : ["alternative_name_server_config", 0, "target_name_servers", 0, "ipv4_address"]
values shouldn't be empty, unless you're trying to whitelist empty space/null to block any alt name server configs? but given what is described, you can add in a list of IPs:
values: ["1.1.1.1", "2.3.2.1", etc...]
whitelist
OR
values: ["1[.]1[.]1[.]*", [[1, 2, 3, 4, 5, 6, 7, 10]]] # this allows the addresses: 1.1.1.1, 1.1.1.2, 1.1.1.3, ..., 1.1.1.7 and 1.1.1.10 but not 1.1.1.8, 1.1.1.9 as 8, 9 are not in the pattern list
pattern whitelist

target_name_server array value is 1- have added a whitelist value for ip address
________________________________
On policies/gcp/clouddns/dns_policy/networks/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096917527>:
correct attribute referencing, but the values to be whitelisted are incorrect. here you want to only allow certain network urls, such as "http://love_writing_polices_for_yotham.net.id<http://love_writing_polices_for_yotham.net.id/>" (note the http is also not secure). you can either:

  1.  hardcode whitelisting:
"values": ["https://love_writing_polices_for_yotham.net.id"],
"policy_type" : "whitelist"

OR even better 2) patterns:
"values": ["https://[.]", [["love_writing_polices_for_yotham"], ["net.id"]]],
"policy_type" : "pattern whitelist"

updated

Hope this is now all good. 🙂

Cheers

Simon




________________________________
From: jerryd10 ***@***.***>
Sent: Tuesday, May 20, 2025 3:22 PM
To: Hardhat-Enterprises/Policy-Deployment-Engine ***@***.***>
Cc: SIMON RANKMORE ***@***.***>; Author ***@***.***>
Subject: Re: [Hardhat-Enterprises/Policy-Deployment-Engine] Gcp/clouddns/dns policy (PR #12)


@jerryd10 requested changes on this pull request.

2nd review round.

Policies very much improved, but you need to configure the attribute paths correctly, and include some values too. Sure these values lists can be configured later but you should add example values of what these later vals can look like.

Iirevalent policies have been removed which is good. the current policies are very much security-related, mostly with allowed dest networks. Good on ya.

________________________________

On policies/gcp/clouddns/dns_managed_zone/auto_create_network/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096945668>:

same as above

________________________________

On policies/gcp/clouddns/dns_managed_zone/cloud_logging/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096943739>:

associated plan.json is missing so i have no point of reference, but from the included c.tf, shouldnt the attribute path be instead ["cloud_logging_config", 0, "enable_logging"]. the value is corrrect.

________________________________

On policies/gcp/clouddns/dns_managed_zone/managed_zone_basic/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096940620>:

Idk what this policy is supposed to do, as it is described to check the dns namespace but an irrelevant attribute is referenced

________________________________

On policies/gcp/clouddns/dns_managed_zone/managed_zone_private/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096939228>:

"private_visibility_config", 0, "networks", 0 "network_url"] and values: [pde.coding_for_yotham.net] as you are now evaluating the private_visibility_config attribute rather than the visibility

________________________________

On policies/gcp/clouddns/dns_managed_zone/private_forwarding/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096936120>:

attribute should be ["forwarding_config", 0, "target_name_servers", 0, "ipv4_address"],
values: ["139.0.0.1"]

________________________________

On policies/gcp/clouddns/dns_managed_zone/private_gke/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096932060>:

shouldn't it be ["private_visibility_config", 0, "gke_cluster"]? I see that you leave the values [] empty to mean that it can be configured later, but you should add some examples in the values [] array to demonstrate how those values added later would look like.

________________________________

On policies/gcp/clouddns/dns_managed_zone/service_directory/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096928577>:

attribute should be ["service_directory_config", 0, "namespace", 0, "namespace_url"]

and you're whitelisting an empty list, so the policy prevents a service dir from being configured?

________________________________

On policies/gcp/clouddns/dns_policy/alt_name_server/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096924941>:

attribute should be:
"attribute_path" : ["alternative_name_server_config", 0, "target_name_servers", 0, "ipv4_address"]
values shouldn't be empty, unless you're trying to whitelist empty space/null to block any alt name server configs? but given what is described, you can add in a list of IPs:

values: ["1.1.1.1", "2.3.2.1", etc...]
whitelist

OR
values: ["1[.]1[.]1[.]*", [[1, 2, 3, 4, 5, 6, 7, 10]]] # this allows the addresses: 1.1.1.1, 1.1.1.2, 1.1.1.3, ..., 1.1.1.7 and 1.1.1.10 but not 1.1.1.8, 1.1.1.9 as 8, 9 are not in the pattern list
pattern whitelist

________________________________

On policies/gcp/clouddns/dns_policy/networks/policy.rego<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#discussion_r2096917527>:

correct attribute referencing, but the values to be whitelisted are incorrect. here you want to only allow certain network urls, such as "http://love_writing_polices_for_yotham.net.id" (note the http is also not secure). you can either:

  1.  hardcode whitelisting:
"values": ["https://love_writing_polices_for_yotham.net.id"],
"policy_type" : "whitelist"

OR even better 2) patterns:
"values": ["https://[.]", [["love_writing_polices_for_yotham"], ["net.id"]]],
"policy_type" : "pattern whitelist"

—
Reply to this email directly, view it on GitHub<https://github.com/Hardhat-Enterprises/Policy-Deployment-Engine/pull/12#pullrequestreview-2852608961>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BHJOKWQ5LF5RBJD5Z4T6MDT27K3ZBAVCNFSM6AAAAAB4PY67HOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDQNJSGYYDQOJWGE>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Cluel355 committed:
updated 21/5/25

Pat-Stuart commented:
Closing to be redone
